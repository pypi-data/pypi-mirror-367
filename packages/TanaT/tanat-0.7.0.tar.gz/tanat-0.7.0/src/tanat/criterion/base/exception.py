#!/usr/bin/env python3
"""
Criterion exceptions.
"""

from ...exception import TanatException


class CriterionException(TanatException):
    """
    Base exception for all criterion-related errors in the TanaT framework.
    """


class InvalidCriterionError(CriterionException):
    """
    Raised when an invalid or improperly configured criterion is encountered.
    """
