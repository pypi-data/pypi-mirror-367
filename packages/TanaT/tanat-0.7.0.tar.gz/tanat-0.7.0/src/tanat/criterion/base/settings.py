#!/usr/bin/env python3
"""
Base class for criterion.
"""

import dataclasses

from .enum import CriterionLevel


@dataclasses.dataclass
class Criterion:
    """
    Base class for criterion.
    """

    _REGISTER_NAME = None

    def get_compatibility_levels(self, max_level=CriterionLevel.TRAJECTORY):
        """List the compatibility levels of the criterion.
        Compatibility levels are marked with decorators.

        Args:
            max_level (CriterionLevel, optional): The maximum level to consider.
                Defaults to Trajectory. Entity < Sequence < Trajectory

        Returns a list of CriterionLevel representing the compatibility levels of the
        criterion. The levels are "sequence","entity" or "trajectory".
        """
        if isinstance(max_level, str):
            max_level = CriterionLevel.from_str(max_level)

        compatibility_levels = getattr(self, "COMPATIBILITY_LEVELS", set())
        return [level for level in compatibility_levels if level <= max_level]
