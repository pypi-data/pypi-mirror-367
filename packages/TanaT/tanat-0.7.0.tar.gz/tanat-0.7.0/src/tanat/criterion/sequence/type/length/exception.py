#!/usr/bin/env python3
"""
Length criterion exception.
"""

from ....base.exception import CriterionException


class ContradictoryLengthCriterionError(CriterionException):
    """
    Raised when length-based filtering criterion are logically incompatible.

    This exception occurs during length criterion configuration when:
    - Attempting to use both 'gt' and 'ge' operators simultaneously
    - Attempting to use both 'lt' and 'le' operators simultaneously
    - Specifying lower and upper bounds that create an impossible condition
      (e.g., lower bound is greater than or equal to upper bound)

    Prevents silent failures in sequence length filtering by explicitly
    highlighting configuration errors.
    """
