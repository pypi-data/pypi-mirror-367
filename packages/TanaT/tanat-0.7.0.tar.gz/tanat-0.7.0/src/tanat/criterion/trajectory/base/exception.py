#!/usr/bin/env python3
"""
Trajectory Criterion exceptions.
"""

from pypassist.mixin.registrable.exceptions import UnregisteredTypeError

from ...base.exception import CriterionException


class TrajectoryCriterionException(CriterionException):
    """
    Trajectory criterion exceptions.
    """


class UnregisteredTrajectoryCriterionTypeError(
    UnregisteredTypeError, TrajectoryCriterionException
):
    """
    Unregistered trajectory criterion type error.
    """
