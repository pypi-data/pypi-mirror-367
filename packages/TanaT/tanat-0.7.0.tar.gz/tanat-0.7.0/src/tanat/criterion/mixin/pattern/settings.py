#!/usr/bin/env python3
"""
Pattern criterion.
"""

from typing import Union

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer
from pypassist.fallback.typing import Dict, List

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible
from .enum import PatternOperator


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class PatternCriterion(Criterion):
    """
    Flexible pattern-based filtering for sequences and entities.

    Enables complex pattern matching across data columns with advanced
    configuration options.

    Args:
        pattern (Dict[str, Union[str, List[str]]]):
            Column patterns to match. Supports:
            - Plain strings
            - Lists of strings
            - Regex patterns (prefix with 'regex:')

        contains (bool, optional):
            - True: Include sequences matching pattern
            - False: Exclude sequences matching pattern
            Defaults to True.

        case_sensitive (bool, optional):
            Whether matching is case-sensitive. Defaults to True.

        operator (PatternOperator, optional):
            Logical operator between patterns:
            - AND: All patterns must match
            - OR: At least one pattern matches
            Defaults to AND.

    Example:
        # Find sequences with emergency events
        PatternCriterion(pattern={'event_type': 'EMERGENCY'})
    """

    pattern: Dict[str, Union[str, List[str]]]
    contains: bool = True
    case_sensitive: bool = True
    operator: PatternOperator = PatternOperator.AND

    _REGISTER_NAME = "pattern"
