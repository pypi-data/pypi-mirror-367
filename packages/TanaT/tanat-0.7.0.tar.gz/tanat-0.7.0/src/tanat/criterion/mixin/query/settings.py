#!/usr/bin/env python3
"""
Settings for query criterion.
"""

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class QueryCriterion(Criterion):
    """
    Pandas-based filtering criterion for sequences and entities.

    Allows complex data filtering using pandas query syntax. Compatible with
    sequence and entity levels, enabling flexible data selection based on
    column conditions.

    Args:
        query (str): Pandas-style query string to filter data.
            Supports complex boolean conditions across columns.

    Example:
        # Filter sequences where age > 50 and gender is 'M'
        criterion = QueryCriterion(query="age > 50 and gender == 'M'")
    """

    query: str

    _REGISTER_NAME = "query"
