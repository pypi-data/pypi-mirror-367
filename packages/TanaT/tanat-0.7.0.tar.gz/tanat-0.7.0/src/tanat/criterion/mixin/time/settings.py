#!/usr/bin/env python3
"""
Settings for time-based filtering.
"""

from typing import Optional, Union
from datetime import datetime

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ...base.settings import Criterion
from ...decorator.entity import entity_compatible
from ...decorator.sequence import sequence_compatible


@viewer
@entity_compatible
@sequence_compatible
@dataclass
class TimeCriterion(Criterion):
    """
    Flexible time-based filtering for sequences and entities.

    Enables precise temporal filtering with multiple time constraints and
    different containment modes depending on the filtering level.

    Args:
        start_before (datetime, optional):
            Filter sequences or entities starting before this timestamp.
        start_after (datetime, optional):
            Filter sequences or entities starting after this timestamp.
        end_before (datetime, optional):
            Filter sequences or entities ending before this timestamp.
        end_after (datetime, optional):
            Filter sequences or entities ending after this timestamp.
        duration_within (bool, optional):
            For entity-level filtering on state/interval entities:
            - True: Entity must be entirely contained within the time range
            - False: Entity can partially overlap with the time range
            Defaults to False. Has no effect on event entities.
        sequence_within (bool, optional):
            For sequence-level filtering (event, interval, state):
            - True: Entire sequence must be contained within the time range
            - False: Sequence can partially overlap with the time range
            Defaults to False.
        date_format (str, optional):
            Custom datetime parsing format for string inputs.

    Examples:
        # Filter entities with partial overlap (default behavior)
        TimeCriterion(
            start_after=datetime.now(),
            end_before=datetime.now() + timedelta(days=90)
        )

        # Only keep state/interval entities fully contained in time range
        TimeCriterion(
            start_after=datetime(2024, 1, 1),
            end_before=datetime(2024, 12, 31),
            duration_within=True
        )

        # Only keep sequences entirely within time range
        TimeCriterion(
            start_after=datetime(2024, 6, 1),
            end_before=datetime(2024, 6, 30),
            sequence_within=True
        )

        # Combine both strict modes
        TimeCriterion(
            start_after=datetime(2024, 1, 1),
            end_before=datetime(2024, 12, 31),
            duration_within=True,
            sequence_within=True
        )
    """

    start_before: Optional[Union[datetime, str, int]] = None
    start_after: Optional[Union[datetime, str, int]] = None
    end_before: Optional[Union[datetime, str, int]] = None
    end_after: Optional[Union[datetime, str, int]] = None
    duration_within: bool = False
    sequence_within: bool = False
    date_format: Optional[str] = None
    # TODO: add settings for duration

    _REGISTER_NAME = "time"
