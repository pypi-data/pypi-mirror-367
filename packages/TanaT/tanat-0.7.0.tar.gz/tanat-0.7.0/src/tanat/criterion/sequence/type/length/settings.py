#!/usr/bin/env python3
"""
Length criterion.
"""

from typing import Optional

from pydantic.dataclasses import dataclass
from pypassist.dataclass.decorators.viewer.decorator import viewer

from ....base.settings import Criterion
from ....decorator.sequence import sequence_compatible
from .exception import ContradictoryLengthCriterionError


@viewer
@sequence_compatible
@dataclass
class LengthCriterion(Criterion):
    """
    Filter sequences based on their length.

    Provides flexible length-based filtering with multiple comparison
    operators for sequence selection.

    Args:
        gt (int, optional):
            Keep sequences with length strictly greater than this value.
        ge (int, optional):
            Keep sequences with length greater than or equal to this value.
        lt (int, optional):
            Keep sequences with length strictly less than this value.
        le (int, optional):
            Keep sequences with length less than or equal to this value.

    Example:
        # Filter sequences with 10 or fewer events
        LengthCriterion(le=10)
    """

    gt: Optional[int] = None  # greater than
    ge: Optional[int] = None  # greater than or equal to
    lt: Optional[int] = None  # less than
    le: Optional[int] = None  # less than or equal to

    _REGISTER_NAME = "length"

    def __post_init__(self):
        if self.gt is not None and self.ge is not None:
            raise ContradictoryLengthCriterionError(
                "Cannot use 'gt' and 'ge' operators simultaneously"
            )
        if self.lt is not None and self.le is not None:
            raise ContradictoryLengthCriterionError(
                "Cannot use 'lt' and 'le' operators simultaneously"
            )

        lower = self.gt if self.gt is not None else self.ge
        upper = self.lt if self.lt is not None else self.le
        if lower is not None and upper is not None and lower >= upper:
            raise ContradictoryLengthCriterionError(
                f"lower bound ({lower}) must be strictly less than upper bound ({upper})."
            )
