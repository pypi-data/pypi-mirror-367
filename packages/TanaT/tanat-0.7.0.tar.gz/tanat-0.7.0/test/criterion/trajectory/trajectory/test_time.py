#!/usr/bin/env python3
"""
Test time criterion applied to single trajectories.
"""

import pytest

from tanat.criterion.mixin.time.settings import TimeCriterion
from tanat.criterion.base.exception import InvalidCriterionError


class TestTimeCriterion:
    """
    Test TimeCriterion applied to single trajectory.
    """

    @pytest.fixture
    def single_trajectory(self, trajectory_pool):
        """
        Return a single trajectory from the pool for testing purposes.
        """
        return trajectory_pool[3].copy()

    @pytest.mark.parametrize(
        "sequence_name",
        ["event", "interval"],
    )
    def test_time_criterion_filter_single_trajectory(
        self, single_trajectory, sequence_name, snapshot
    ):
        """
        Test filtering a single trajectory using TimeCriterion with `inplace=False`.
        """
        time_criterion = TimeCriterion(start_before="2023-04-30")

        filtered_trajectory = single_trajectory.filter(
            criterion=time_criterion,
            sequence_name=sequence_name,
            inplace=False,
        )

        filtered_sequence_data = filtered_trajectory[sequence_name].sequence_data
        snapshot.assert_match(filtered_sequence_data.to_csv())

    @pytest.mark.parametrize(
        "sequence_name",
        ["event", "interval"],
    )
    def test_time_criterion_filter_single_trajectory_inplace(
        self, single_trajectory, sequence_name, snapshot
    ):
        """
        Test filtering a single trajectory using TimeCriterion with `inplace=True`.
        """
        time_criterion = TimeCriterion(start_before="2023-04-30")

        result = single_trajectory.filter(
            criterion=time_criterion,
            sequence_name=sequence_name,
            inplace=True,
        )

        assert result is None

        modified_sequence_data = single_trajectory[sequence_name].sequence_data
        snapshot.assert_match(modified_sequence_data.to_csv())

    @pytest.mark.parametrize(
        "sequence_name",
        ["event", "interval"],
    )
    def test_time_criterion_filter_consistency_between_inplace_and_copy(
        self, single_trajectory, sequence_name
    ):
        """
        Ensure consistent filtering results between inplace and non-inplace TimeCriterion filtering.
        """
        time_criterion = TimeCriterion(start_before="2023-04-30")

        # Non-inplace
        filtered_trajectory = single_trajectory.filter(
            criterion=time_criterion,
            sequence_name=sequence_name,
            inplace=False,
        )
        non_inplace_result = filtered_trajectory[sequence_name].sequence_data

        # Inplace
        trajectory_copy = single_trajectory.copy()
        trajectory_copy.filter(
            criterion=time_criterion,
            sequence_name=sequence_name,
            inplace=True,
        )
        inplace_result = trajectory_copy[sequence_name].sequence_data

        assert non_inplace_result.equals(inplace_result)

    def test_time_criterion_filter_state_sequence_raises(self, single_trajectory):
        """
        Test that applying TimeCriterion to the 'state' sequence raises NotImplementedError.
        """
        time_criterion = TimeCriterion(start_before="2023-04-30")

        with pytest.raises(NotImplementedError):
            single_trajectory.filter(
                criterion=time_criterion,
                sequence_name="state",
                inplace=False,
            )

        trajectory_copy = single_trajectory.copy()
        with pytest.raises(NotImplementedError):
            trajectory_copy.filter(
                criterion=time_criterion,
                sequence_name="state",
                inplace=True,
            )

    @pytest.mark.parametrize(
        "sequence_name",
        ["event", "interval"],
    )
    def test_time_criterion_dict_equivalence(self, single_trajectory, sequence_name):
        """
        Validate that using a dict-based time criterion yields
        the same result as using a TimeCriterion object.
        """
        dict_criterion = {"start_before": "2023-04-30"}

        filtered_from_dict = single_trajectory.filter(
            criterion=dict_criterion,
            criterion_type="time",
            sequence_name=sequence_name,
            inplace=False,
        )

        time_criterion_obj = TimeCriterion(start_before="2023-04-30")
        filtered_from_obj = single_trajectory.filter(
            criterion=time_criterion_obj,
            sequence_name=sequence_name,
            inplace=False,
        )

        assert filtered_from_dict[sequence_name].sequence_data.equals(
            filtered_from_obj[sequence_name].sequence_data
        )

    def test_time_criterion_filter_preserves_other_sequences(self, single_trajectory):
        """
        Ensure that applying TimeCriterion to one sequence does not alter unrelated sequences.
        """
        time_criterion = TimeCriterion(start_before="2023-04-30")

        filtered_trajectory = single_trajectory.filter(
            criterion=time_criterion,
            sequence_name="event",
            inplace=False,
        )

        for sequence_name in single_trajectory.sequences:
            if sequence_name != "event":
                assert filtered_trajectory[sequence_name].sequence_data.equals(
                    single_trajectory[sequence_name].sequence_data
                )

    @pytest.mark.parametrize(
        "sequence_name",
        ["event", "interval"],
    )
    def test_time_criterion_filter_with_no_matching_elements_returns_empty(
        self, single_trajectory, sequence_name
    ):
        """
        Test that applying TimeCriterion with no matches returns an empty sequence.

        Uses an unrealistic date far in the future to ensure no records match.
        """
        time_criterion = TimeCriterion(start_after="2100-12-12")

        filtered_trajectory = single_trajectory.filter(
            criterion=time_criterion,
            sequence_name=sequence_name,
            inplace=False,
        )

        assert filtered_trajectory.sequences.get(sequence_name) is None

    def test_match_method_raises_error(self, single_trajectory):
        """
        Test match method raises an error.
        Match method is only applicable for 'trajectory' compatible criterion.
        """
        criterion = TimeCriterion(start_before="2023-04-30")

        with pytest.raises(InvalidCriterionError):
            single_trajectory.match(criterion=criterion)

        ## -- from dict
        with pytest.raises(InvalidCriterionError):
            single_trajectory.match(
                criterion={"start_before": "2023-04-30"},
                criterion_type="time",
            )
