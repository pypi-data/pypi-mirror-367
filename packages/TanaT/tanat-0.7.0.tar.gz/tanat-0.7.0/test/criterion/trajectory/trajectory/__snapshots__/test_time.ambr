# serializer version: 1
# name: TestTimeCriterion.test_time_criterion_filter_single_trajectory[event]
  '''
  patient_id,date,event_type,provider
  3,2023-01-28,CONSULTATION,RADIOLOGIST
  3,2023-02-02,FOLLOW_UP,NURSE
  3,2023-02-10,FOLLOW_UP,LABORATORY
  3,2023-03-06,CONSULTATION,
  3,2023-03-22,EMERGENCY,
  
  '''
# ---
# name: TestTimeCriterion.test_time_criterion_filter_single_trajectory[interval]
  '''
  patient_id,start_date,end_date,medication,administration_route,dosage
  3,2023-04-15,2023-04-24,ANTIBIOTIC,INHALATION,3
  
  '''
# ---
# name: TestTimeCriterion.test_time_criterion_filter_single_trajectory_inplace[event]
  '''
  patient_id,date,event_type,provider
  3,2023-01-28,CONSULTATION,RADIOLOGIST
  3,2023-02-02,FOLLOW_UP,NURSE
  3,2023-02-10,FOLLOW_UP,LABORATORY
  3,2023-03-06,CONSULTATION,
  3,2023-03-22,EMERGENCY,
  
  '''
# ---
# name: TestTimeCriterion.test_time_criterion_filter_single_trajectory_inplace[interval]
  '''
  patient_id,start_date,end_date,medication,administration_route,dosage
  3,2023-04-15,2023-04-24,ANTIBIOTIC,INHALATION,3
  
  '''
# ---
